db.go

```diff
@@ -423,17 +423,27 @@ func (db *DB) mmap(minsz int) error {
 	}
 
 	// Unmap existing data before continuing.
-	if err := db.munmap(); err != nil {
+	if err = db.munmap(); err != nil {
 		return err
 	}
 
 	// Memory-map the data file as a byte slice.
 	// gofail: var mapError string
 	// return errors.New(mapError)
-	if err := mmap(db, size); err != nil {
+	if err = mmap(db, size); err != nil {
 		return err
 	}
 
+	// Perform unmmap on any error to reset all data fields:
+	// dataref, data, datasz, meta0 and meta1.
+	defer func() {
+		if err != nil {
+			if unmapErr := db.munmap(); unmapErr != nil {
+				err = fmt.Errorf("%w; unmap failed: %v", err, unmapErr)
```

Comment:
@ahrtr 

1. I would be explicit in that error that it's 'rollback unmap also failed'.

2. I had some doubts whether we should execute here unmap or just db.invalidate. 
Ideally cleanup for failing mmap() (like rollback unmap) should be handled within the mmap() function's logic, rather than assumed externally. 

3. If we suspect the failing rollbacks might be reason for the double-linking issues, 
by submitting this patch, we are making it more difficult to build repro (i.e. we need to test with & without that patch). It would be good to split failpoints (helping with repro) from the 'fix' PR. 

4. Next time please ping me or @serathius a little longer for business logic change - to not create precedences.

Reply:
> 1. I would be explicit in that error that it's 'rollback unmap also failed'.

Thanks for the suggestion. Will fix it separately.

> 
> 2. I had some doubts whether we should execute here unmap or just db.invalidate.
> Ideally cleanup for failing mmap() (like rollback unmap) should be handled within the mmap() function's logic, rather than assumed externally.


Definitely we should execute `db.unmap` instead of `db.invalidate`. Note that the deferred function will only be executed when mmap already succeeded but (1) mlock somehow failed or (2) both meta pages are corrupted.

 

> 3. If we suspect the failing rollbacks might be reason for the double-linking issues,
> by submitting this patch, we are making it more difficult to build repro (i.e. we need to test with & without that patch). It would be good to split failpoints (helping with repro) from the 'fix' PR.


Not sure what do you mean by "double-linking issues", could you clarify this?  I found the issue with the new added test cases in this PR, so I fixed it in the same PR. I plan to backport both failpoints and all bug fixes to release-1.3, please see https://github.com/etcd-io/bbolt/issues/436. Please let me know if you still have any concern.

> 4. Next time please ping me or @serathius a little longer for business logic change - to not create precedences.

Sorry to merge this PR without another maintainer's approval. I do think it's a safe change, and It's already reviewed by three experienced contributors. But anyway, will try to follow the process next time.

Reply:
1. thx
2. You are right. 
3. I mean corruptions like #402. There is slim chance there are caused by transaction rollback and follow up commit.  If we find repro without this PR, we have a higher confidence we found the root issue. 
If the PR indeed fixes the problem. You might struggle to repro problem that got already fixed. 
I'm not saying I have concerns about this PR, just flagging.
4. Thx.

Reply:
> If the PR indeed fixes the problem. You might struggle to repro problem that got already fixed.
> I'm not saying I have concerns about this PR, just flagging.

Thanks for the reminder.